pkg/tanka/find.go

```diff
@@ -2,123 +2,165 @@ package tanka
+
+	// find all jsonnet files within given paths
+	pathChan := make(chan string, len(paths))
+	findJsonnetFilesChan := make(chan findJsonnetFilesOut)
+	for i := 0; i < opts.Parallelism; i++ {
+		go func() {
+			for path := range pathChan {
+				jsonnetFiles, err := jsonnet.FindFiles(path, nil)
+				var mainFiles []string
+				for _, file := range jsonnetFiles {
+					if filepath.Base(file) == jpath.DefaultEntrypoint {
+						mainFiles = append(mainFiles, file)
+					}
+				}
+				findJsonnetFilesChan <- findJsonnetFilesOut{jsonnetFiles: mainFiles, err: err}
+			}
+		}()
 	}
 
-	// if path is a file, don't search deeper
-	if !stat.IsDir() {
-		return nil, nil
+	// push paths to channel
+	var pathMap = map[string]bool{} // prevent duplicates
+	for _, path := range paths {
+		if _, ok := pathMap[path]; ok {
+			continue
+		}
+		pathMap[path] = true
+		pathChan <- path
 	}
 
-	// list directory
-	files, err := os.ReadDir(path)
-	if err != nil {
-		return nil, findErr(path, err)
+	// collect jsonnet files
+	var jsonnetFiles []string
+	for i := 0; i < len(paths); i++ {
+		res := <-findJsonnetFilesChan
+		if res.err != nil {
+			return nil, res.err
+		}
+		jsonnetFiles = append(jsonnetFiles, res.jsonnetFiles...)
+	}
+	close(pathChan)
+	close(findJsonnetFilesChan)
+
+	findJsonnetFilesEndTime := time.Now()
+
+	// find all environments within jsonnet files
+	jsonnetFilesChan := make(chan string, len(jsonnetFiles))
+	findEnvsChan := make(chan findEnvsOut)
+
+	for i := 0; i < opts.Parallelism; i++ {
+		go func() {
+			for jsonnetFile := range jsonnetFilesChan {
+				// try if this has envs
+				list, err := List(jsonnetFile, Opts{JsonnetOpts: opts.JsonnetOpts})
+				if err != nil &&
+					// expected when looking for environments
+					!errors.As(err, &jpath.ErrorNoBase{}) &&
+					!errors.As(err, &jpath.ErrorFileNotFound{}) {
+					findEnvsChan <- findEnvsOut{err: fmt.Errorf("%s:\n %w", jsonnetFile, err)}
+					continue
+				}
+				filtered := []*v1alpha1.Environment{}
+				// optionally filter
+				if opts.Selector != nil && !opts.Selector.Empty() {
+					for _, e := range list {
+						if !opts.Selector.Matches(e.Metadata) {
+							continue
+						}
+						filtered = append(filtered, e)
+					}
+				} else {
+					filtered = append(filtered, list...)
+				}
+				findEnvsChan <- findEnvsOut{envs: filtered, err: nil}
+			}
+		}()
 	}
 
-	// it's not one. Maybe subdirectories are?
-	ch := make(chan findOut)
-	routines := 0
-
-	// recursively find in parallel
-	for _, fi := range files {
-		if !fi.IsDir() {
+	// push jsonnet files to channel
+	var jsonnetFileMap = map[string]bool{} // prevent duplicates
+	for _, jsonnetFile := range jsonnetFiles {
+		if _, ok := jsonnetFileMap[jsonnetFile]; ok {
 			continue
 		}
-
-		routines++
-		go findShim(filepath.Join(path, fi.Name()), opts, ch)
+		jsonnetFileMap[jsonnetFile] = true
+		jsonnetFilesChan <- jsonnetFile
 	}
 
-	// collect parallel results
-	var errs []error
+	// collect environments
```

Comment:
close the channel here instead of a few lines down - this records no more values will be sent